ZelBack/src/services/bandwidthMonitoringService.js

```diff
@@ -0,0 +1,298 @@
+ * @returns {Promise<number>} Speed in Mbps
+ */
+async function getNodeBandwidth() {
+  try {
+    const benchmarks = await benchmarkService.getBenchmarks();
+    if (benchmarks.status === 'success' && benchmarks.data) {
+      // Benchmark returns download speed in Mbps
+      const downloadSpeed = benchmarks.data.download || 0;
+      return downloadSpeed;
+    }
+    return 100; // Default value if unable to obtain
+  } catch (error) {
+    log.error('Error getting bandwidth benchmark:', error);
+    return 100; // Default value
+  }
+}
+
+/**
+ * Get network statistics from a Docker container
+ * @param {string} containerName Container name
+ * @returns {Promise<object>} Network statistics
+ */
+async function getContainerNetworkStats(containerName) {
+  try {
+    const container = await dockerService.getDockerContainerByIdOrName(containerName);
+    const stats = await container.stats({ stream: false });
+    
+    // Calculate current bandwidth
+    const networks = stats.networks || {};
+    let totalRxBytes = 0;
+    let totalTxBytes = 0;
+    
+    for (const network of Object.values(networks)) {
+      totalRxBytes += network.rx_bytes || 0;
+      totalTxBytes += network.tx_bytes || 0;
+    }
+    
+    return {
+      timestamp: Date.now(),
+      rxBytes: totalRxBytes,
+      txBytes: totalTxBytes,
+      totalBytes: totalRxBytes + totalTxBytes
+    };
+  } catch (error) {
+    log.error(`Error getting network stats from container ${containerName}:`, error);
+    return null;
+  }
+}
+
+/**
+ * Calculate bandwidth speed in Mbps
+ * @param {object} previousStats Previous statistics
+ * @param {object} currentStats Current statistics
+ * @returns {number} Speed in Mbps
+ */
+function calculateBandwidthSpeed(previousStats, currentStats) {
+  if (!previousStats || !currentStats) return 0;
+  
+  const timeDiff = (currentStats.timestamp - previousStats.timestamp) / 1000; // seconds
+  const bytesDiff = currentStats.totalBytes - previousStats.totalBytes;
+  
+  // Convert bytes/second to Mbps
+  const bytesPerSecond = bytesDiff / timeDiff;
+  const mbps = (bytesPerSecond * 8) / (1024 * 1024);
+  
+  return mbps;
+}
+
+/**
+ * Apply throttling to a Docker container
+ * @param {string} containerName Container name
+ * @param {number} limitMbps Limit in Mbps
+ */
+async function applyThrottling(containerName, limitMbps) {
+  try {
+    // Use tc (traffic control) to limit bandwidth
+    const limitKbps = Math.floor(limitMbps * 1024);
+    
+    // Get container PID
+    const container = await dockerService.getDockerContainerByIdOrName(containerName);
+    const containerInfo = await container.inspect();
+    const pid = containerInfo.State.Pid;
+    
+    // Apply throttling using tc in container's network namespace
+    const commands = [
+      `nsenter -t ${pid} -n tc qdisc del dev eth0 root 2>/dev/null || true`,
+      `nsenter -t ${pid} -n tc qdisc add dev eth0 root handle 1: htb default 30`,
```

Comment:
I'm assuming this has been researched so we know it's best practice and also tested?!?

This is assuming it's running as root. (Which it is on Arcane) but not on legacy. Also we should be using our serviceHelper.runCommand - not this execAsync stuff.

I'd check with Kamil though - he's already added some traffic control stuff to every bridge that gets added...

<img width="1398" height="394" alt="Screenshot 2025-08-04 at 2 16 28 PM" src="https://github.com/user-attachments/assets/cfff1347-39db-40c9-b796-0b683f9d768c" />